src/Raven.Server/Utils/ThreadPool.cs

```diff
@@ -0,0 +1,224 @@
+                _threadFieldName = runtimeThread.GetType().GetField("m_Name", System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.NonPublic);
+                Debug.Assert(_threadFieldName != null);
+                
+            }
+
+            public PooledThread(RavenThreadPool pool)
+            {                
+                _parent = pool;
+                _logger = LoggingSource.Instance.GetLogger<PooledThread>("PooledThread");
+            }
+
+            public LongRunningWork SetWorkForThread(Action<object> action, object state, string name)
+            {
+                _action = action;
+                _state = state;
+                _name = name;
+                _workIsDone = new LongRunningWork(new ManualResetEvent(false));
+
+                _waitForWork.Set();
+                return _workIsDone;
+            }
+
+
+            private void ReduceMemoryUsage()
+            {                              
+                // not sure about all of this "cleaning", I suspect that most of it will actually harm performance and make GC work harder...
+                ByteStringMemoryCache.CleanForCurrentThread();
+
+                using (var docsContextPool = new DocumentsContextPool(null))
+                    docsContextPool.Clean();
+
+                using (var transactionContextPool = new TransactionContextPool(null))                
+                    transactionContextPool.Clean();
+
+                using (var jsonContextPool = new JsonContextPool())
+                    jsonContextPool.Clean();
+                                
+                new Tree(true).CleanLocalBuffer();
+
+                StreamExtensions.CleanBuffer();
+
+                ValueReader.CleanBuffer();
+
+                BlittableWriter<UnmanagedWriteBuffer>.CleanPropertyArrayOffset();
+                BlittableWriter<UnmanagedStreamBuffer>.CleanPropertyArrayOffset();
+
+                LazyStringValue.CleanBuffers();
+
+                CaseInsensitiveStringSegmentEqualityComparer.CleanBuffer();
+
+                BatchRequestParser.CleanCache();
+
+                JsBlittableBridge.CleanCache();
+
+                DocumentIdWorker.CleanCache();
+                Raven.Server.Json.BlittableJsonTextWriterExtensions.CleanCache();
+
+                ChangeVectorUtils.CleanCache();
+
+
+            }
+
+
+
+            public void Run()
+            {
+                while (true)
+                {
+                    ResetCurrentThreadName();
+                    Thread.CurrentThread.Priority = ThreadPriority.Normal;
+                    Thread.CurrentThread.Name = "Available Pool Thread";
+
+                    _waitForWork.WaitOne();
+                    _workIsDone.ManagedThreadId = Thread.CurrentThread.ManagedThreadId;
+                    if (_action == null)
+                        return; // should only happen when we shutdown
+
+                    ResetCurrentThreadName();
```

Comment:
I wonder why do we need to reset it first before using `Thread.CurrentThread.Name = _name;`?

Reply:
It was set to `"Available Pool Thread"` on line 177, and it cannot be set to a value unless it is null

Reply:
Good to know. Thanks